### PR TITLE
Define request element picture in picture algorithm

### PR DESCRIPTION
**Pull request by beccahughes**
- Renames the existing request Picture-in-Picture algorithm to include video
- Splits out the common parts of this algorithm
- Defines a new element request Picture-in-Picture algorithm

***
<a href="https://pr-preview.s3.amazonaws.com/WICG/picture-in-picture/pull/134.html" title="Last updated on May 24, 2019, 8:56 PM UTC (89274f5)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/picture-in-picture/134/e5eaadc...89274f5.html" title="Last updated on May 24, 2019, 8:56 PM UTC (89274f5)">Diff</a>

### CODE REVIEW

**Comment by mounirlamouri** [comment]
@beaufortfrancois can you PTAL? I removed myself from the review list but add me back if you want my take.

File: index.bs [hunk 1]
```diff
@@ -254,37 +322,46 @@ is determined by the platform and will be left to the platform.
...
-9. Set {{pictureInPictureElement}} to |video|.
-10. Let <dfn>Picture-in-Picture window</dfn> be a new instance of
-    {{PictureInPictureWindow}} associated with {{pictureInPictureElement}}.
-11. <a>Queue a task</a> to <a>fire an event</a> with the name
-    {{enterpictureinpicture}} at the |video| with its {{bubbles}} attribute
-    initialized to `true` and its {{pictureInPictureWindow}} attribute
-    initialized to <a>Picture-in-Picture window</a>.
+5. If |video| is {{pictureInPictureElement}}, abort these steps.
+6. If |playingRequired| is `true` and |video| is {{paused}}, abort these steps.
+7. Run the <a>update Picture-in-Picture element algorithm</a> with |video|.
```
Comment (beaufortfrancois) [inline index.bs]: Do we intentionally miss to pass `|Document|`?

File: index.bs [hunk 2]
```diff
@@ -179,8 +190,65 @@ extensible.
...
+
+## Update Picture-in-Picture Element ## {#update-element}
+
+When the <dfn>update Picture-in-Picture element algorithm</dfn> with |element|
+and |document| is invoked, the user agent MUST run the following steps:
+
+1. Set {{pictureInPictureElement}} to |element|.
+2. Let <dfn>Picture-in-Picture window</dfn> be a new instance of
+    {{PictureInPictureWindow}} associated with {{pictureInPictureElement}} with
+    a reference to |document|.
```
Comment (beaufortfrancois) [inline index.bs]: It isn't the case for video. Shall we do this separately?

File: index.bs [hunk 3]
```diff
@@ -179,8 +190,65 @@ extensible.
...
+
+<button id="pipButton"></button>
+
+<script>
+  const video = document.getElementById('video');
+  const pipButton = document.getElementById('pipButton');
+  const interactiveUI = document.getElementById('interactive-ui');
+  const nonInteractiveUI = document.getElementById('non-interactive-ui');
+
+  pipButton.addEventListener('click', () => {
```
[the remark this answers is not in the document]
Comment (beccahughes) [reply index.bs]: How would the website PiP content if there is no button to do so?

File: index.bs (hunk 1 above)
Comment (beccahughes) [reply index.bs]: Yes there is no document for video pip

File: index.bs [hunk 4]
```diff
@@ -254,37 +322,46 @@ is determined by the platform and will be left to the platform.
...
 run the following steps:
 
-1. If <a>Picture-in-Picture support</a> is `false`, throw a
-    {{NotSupportedError}} and abort these steps.
-2. If the document is not allowed to use the <a>policy-controlled feature</a>
-    named `"picture-in-picture"`, throw a {{SecurityError}} and abort these
-    steps.
-3. If |video|'s {{readyState}} attribute is {{HAVE_NOTHING}}, throw a
+1. Run the <a>check Picture-in-Picture is allowed algorithm</a> with |video| and
+    |userActivationRequired|. If an error is thrown then throw this error.
```
[the remark this answers is not in the document]
Comment (beccahughes) [reply index.bs]: Done.

File: index.bs (hunk 2 above)
Comment (beccahughes) [reply index.bs]: Done.

File: index.bs [hunk 5]
```diff
@@ -179,8 +190,65 @@ extensible.
...
+<script>
+  const video = document.getElementById('video');
+  const pipButton = document.getElementById('pipButton');
+  const interactiveUI = document.getElementById('interactive-ui');
+  const nonInteractiveUI = document.getElementById('non-interactive-ui');
+
+  pipButton.addEventListener('click', () => {
+    interactiveUI.requestPictureInPicture({ interactive: true })
+      .catch(e =>
+          return nonInteractiveUI.requestPictureInPicture({ interactive: false }))
```
[the remark this answers is not in the document]
Comment (beccahughes) [reply index.bs]: Done.

File: index.bs [hunk 6]
```diff
@@ -179,8 +190,65 @@ extensible.
...
+      .then(e =>
+          e.pictureInPictureWindow.document.body.appendChild(video));
```
[the remark this answers is not in the document]
Comment (beccahughes) [reply index.bs]: Done.

File: index.bs [hunk 7]
```diff
@@ -179,8 +190,65 @@ extensible.
...
+the following steps:
+
+1. If <a>Picture-in-Picture support</a> is `false`, throw a
+    {{NotSupportedError}} and abort these steps.
+2. If the document is not allowed to use the <a>policy-controlled feature</a>
+    named `"picture-in-picture"`, throw a {{SecurityError}} and abort these
+    steps.
+3. If |userActivationRequired| is `true` and the algorithm is not
+    <a>triggered by user activation</a>, throw a {{NotAllowedError}} and
+    abort these steps.
```
[the remark this answers is not in the document]
Comment (beccahughes) [reply index.bs]: Done.

File: index.bs [hunk 8]
```diff
@@ -179,8 +190,65 @@ extensible.
...
+When the <dfn>check Picture-in-Picture is allowed algorithm</dfn> with
+|element| and |userActivationRequired| is invoked, the user agent MUST run
+the following steps:
+
+1. If <a>Picture-in-Picture support</a> is `false`, throw a
+    {{NotSupportedError}} and abort these steps.
+2. If the document is not allowed to use the <a>policy-controlled feature</a>
+    named `"picture-in-picture"`, throw a {{SecurityError}} and abort these
+    steps.
+3. If |userActivationRequired| is `true` and the algorithm is not
```
[the remark this answers is not in the document]
Comment (beccahughes) [reply index.bs]: Done.

File: index.bs (hunk 3 above)
[the remark this answers is not in the document]
Comment (mounirlamouri) [reply index.bs]: I think Francois' point is that a website may want to only have interactive Picture-in-Picture and they would want to do early detection. I think that's a fair point and maybe it would require adding something to the API? I would suggest to open an issue about this instead of blocking this PR though.

[the remark this answers is not in the document]
Comment (beccahughes) [reply index.bs]: Ah I see. I have filed https://github.com/WICG/picture-in-picture/issues/139

File: index.bs [hunk 9]
```diff
@@ -179,55 +190,64 @@ extensible.
...
+
+  pipButton.addEventListener('click', () => {
+    interactiveUI.requestPictureInPicture({ interactive: true })
+      .catch(e =>
+          return nonInteractiveUI.requestPictureInPicture({ interactive: false }))
+      .then(e => {
+          e.pictureInPictureWindow.document.body.appendChild(video);
+
+          // TODO: Setup JavaScript and CSS for Picture-in-Picture window.
+      });
```
Comment (beaufortfrancois) [inline index.bs]: Remove `;`

File: index.bs [hunk 10]
```diff
@@ -179,55 +190,64 @@ extensible.
...
+<script>
+  const video = document.getElementById('video');
+  const pipButton = document.getElementById('pipButton');
+  const interactiveUI = document.getElementById('interactive-ui');
+  const nonInteractiveUI = document.getElementById('non-interactive-ui');
+
+  pipButton.addEventListener('click', () => {
+    interactiveUI.requestPictureInPicture({ interactive: true })
+      .catch(e =>
+          return nonInteractiveUI.requestPictureInPicture({ interactive: false }))
```
Comment (beaufortfrancois) [inline index.bs]: Remove `return` or add `{..}`.
I prefer removing `return` ;)

File: index.bs (hunk 9 above)
Comment (beccahughes) [reply index.bs]: Done

File: index.bs (hunk 10 above)
Comment (beccahughes) [reply index.bs]: Done

File: index.bs [hunk 11]
```diff
@@ -252,6 +272,125 @@ pause).
...
+3. If the algorithm is not <a>triggered by user activation</a>, throw a
+    {{NotAllowedError}} and abort these steps.
+4. If |element| is {{pictureInPictureElement}}, abort these steps.
+6. Let |pictureInPictureBrowsingContext| be a new <a>browsing context</a> with the
+    <a>opener browsing context</a> set to the current document's browsing context.
+7. Let |document| be a new {{Document}} with its browsing context set to
+    |pictureInPictureBrowsingContext| and the URL set to <a>about:blank</a>.
+8. Let |clone children flag| be `true`.
+9. Run <a>node clone</a> with |element|, |document| and |clone children flag|.
+10. Let <a>Picture-in-Picture window</a> be a new instance of
```
Comment (beaufortfrancois) [inline index.bs]: You may want to update this text as well in #pictureinpicturewindow.
```A PictureInPictureWindow instance represents a Picture-in-Picture window associated with an HTMLVideoElement. ```

File: index.bs [hunk 12]
```diff
@@ -252,6 +272,125 @@ pause).
...
+
+# Video Specific Concepts # {#video-concepts}
+
+## Request Video Picture-in-Picture ## {#request-video-pip}
+
+When the <dfn>request video Picture-in-Picture algorithm</dfn> with |video|,
+|userActivationRequired| and |playingRequired| is invoked, the user agent MUST
+run the following steps:
+
+1. Run the <a>check Picture-in-Picture is allowed algorithm</a> with |video|.
```
Comment (beaufortfrancois) [inline index.bs]: I don't think you need to pass `|video|` or `|element|` for this algorithm.

File: index.bs [hunk 13]
```diff
@@ -252,6 +272,125 @@ pause).
...
+4. If |element| is {{pictureInPictureElement}}, abort these steps.
+6. Let |pictureInPictureBrowsingContext| be a new <a>browsing context</a> with the
+    <a>opener browsing context</a> set to the current document's browsing context.
+7. Let |document| be a new {{Document}} with its browsing context set to
+    |pictureInPictureBrowsingContext| and the URL set to <a>about:blank</a>.
+8. Let |clone children flag| be `true`.
+9. Run <a>node clone</a> with |element|, |document| and |clone children flag|.
+10. Let <a>Picture-in-Picture window</a> be a new instance of
+    {{PictureInPictureWindow}} associated with |element| and with a reference to
+    |document|.
```
Comment (beaufortfrancois) [inline index.bs]: Should this be part of another PR as there is no `document` yet. WDYT?

File: index.bs (hunk 11 above)
Comment (beccahughes) [reply index.bs]: Done.

File: index.bs (hunk 12 above)
Comment (beccahughes) [reply index.bs]: Done.

File: index.bs (hunk 13 above)
Comment (beccahughes) [reply index.bs]: We define this in https://github.com/WICG/picture-in-picture/pull/136 so we could always wait to submit that PR before this one

**Comment by beaufortfrancois** [comment]
Closing https://github.com/w3c/picture-in-picture/pull/136 makes this PR no-op.
